stdlib/asm/crypto/dsa/rpo_falcon512.masm

```diff
@@ -125,80 +125,69 @@ end
 
-#! For an element `tau := (tau0, tau1)` in the quadratic extension field, computes all its powers
-#! `tau^i` for `i = 0,..., 512` and stores them in the memory region `[tau_ptr, tau_ptr + 513*4)`.
-#! The procedure returns `tau_ptr + 513*4`.
-#!
-#! Input: [tau1, tau0, tau_ptr, ...]
-#! Output: [tau_ptr + 513*4, ...]
-#!
-#! Cycles: ~8900
-export.powers_of_tau
-
-    # 1) Save tau^0 i.e. (0, 1)
-    push.1 push.0.0.0
-    dup.6 add.4 swap.7
-    mem_storew
-    drop drop
-    #=> [0, 1, tau1, tau0, tau_ptr+4, ...]
-
-    # 2) Compute tau^i
-    repeat.512
-        dupw ext2mul
-
-        movup.3 movup.3
-
-        dup.6 add.4 swap.7 mem_storew
-
-        drop drop
-    end
-
-    dropw
-    #=> [tau_ptr + 513*4, ...]
-end
-
-
-#! Sets the memory region `[ptr, ptr + 512*4)` to zero. The pointer c_ptr := ptr + 512*4 is returned
-#! to be used to store the hash-to-point polynomial of the message later on.
-#!
-#! Input: [ptr, ...]
-#! Output: [...]
-#!
-#! Cycles: ~3100
-export.set_to_zero
-    padw
-    repeat.512
-        dup.4 add.4 swap.5
-        mem_storew
-    end
-    dropw
-end
-
 #! Takes as input PK, the hash of the coefficients of the polynomial `h` representing the expanded
 #! public key, and a pointer to the memory location where the coefficients of the polynomial `h`
 #! will be stored.
 #! The procedure loads `h` from the advice stack and compares its hash with the provided hash `PK`.
 #! It then loads the polynomial `s2` representing the signature from the advice stack and lays it
 #! in memory right after `h`.
-#! It then loads the claimed polynomial `h * s2` in Z_Q[x] where Q is the Miden VM prime from
-#! the advice stack and lays it right after `s2`.
+#! After that, it loads the claimed polynomial `h * s2` in Z_Q[x] where Q is the Miden VM prime
+#! from the advice stack and lays it right after `s2`.
 #! The hash of `h`, `s2` and the claimed product is also computed and the first two field elements
-#! of the digest (i.e., the Fiat-Shamir challenge) are returned on the stack alongside
-#! the incremented pointer.
+#! of the digest (i.e., the Fiat-Shamir challenge) are used in order to check that 
+#! pi == h * s2 in Z_Q[x] by evaluating both sides at the random point. Note that since we need
+#! to hash the polynomials before computing their evaluations, the procedure receives
+#! the evaluation point non-deterministically so that it can perform the evaluations while hashing
+#! is still going on. After hashing of the polynomials is finished, the procedure checks that
+#! the evaluation point was indeed derived correctly from the hash of all 3 polynomials.
 #!
 #! Input: [ptr, PK, ...]
-#! Output: [tau1, tau0, ptr + 512*4 ...]
+#! Output: [...]
 #!
-#! Cycles: 5049
-export.load_h_s2_and_product.4
+#! Cycles: 4430
+export.load_h_s2_and_product.16
 
-    # 1) Store PK for later comparison
-    movdn.4
+    # 1) Save pointer to polynomials as we will be using it in several places
```

Comment:
Before we get into the function body itself, it may be helpful to write what we are going to use the 16 local variables for.

Reply:
Added